VirtualMachineExport documentation (#549)
kubevirt/user-guide: docs/operations/export_api.md

# Export API

> 

Reviewer: It would be nice here to have a bit of intro text so the reader can quickly scan from search results to see if this is what they're after.

Also, small nit, you go from H1 straight down to H3, bypassing H2... You could promote most of your headers by one or add some big idea chapter headings like Prerequisites.

Author: Reason I go from H1 to H3 is that H2 renders almost the same as H1 and it is hard to tell the difference between them. I can definitely add some intro text to explain the point of this.

### Export Feature Gate

VMExport support must be enabled in the feature gates to be available. The
[feature gates](./activating_feature_gates.md#how-to-activate-a-feature-gate)
field in the KubeVirt CR must be expanded by adding the `VMExport` to it.

### Export token

> In order to securely export a Virtual Machine Disk, you must create a token that is used to authorize users accessing the export endpoint. This token must be in the namespace the Virtual Machine resides. The contents of the secret can be passed as a token header or parameter to the export URL. The name of the header or argument is `x-kubevirt-export-token` with a value that matches the content of the secret. The secret can be named any valid secret in the namespace. We recommend you generate an alpha numeric token of at least 12 characters. The data key should be `token`. For example:

Reviewer: ```suggestion
In order to securely export a Virtual Machine Disk, you must create a token that is used to authorize users accessing the export endpoint. This token must be in the same namespace as the Virtual Machine. The contents of the secret can be passed as a token header or parameter to the export URL. The name of the header or argument is `x-kubevirt-export-token` with a value that matches the content of the secret. The secret can be named any valid secret in the namespace. We recommend you generate an alpha numeric token of at least 12 characters. The data key should be `token`. For example:
```

stringData:
  token: 1234567890ab
```
### Export Virtual Machine volumes
After you have created the token you can now create a VMExport CR that identifies the Virtual Machine you want to export. You can create a VMExport that looks like this:

```yaml
apiVersion: export.kubevirt.io/v1alpha1
kind: VirtualMachineExport
metadata:
  name: example-export
spec:
  tokenSecretRef: example-token
  source:
    apiGroup: "kubevirt.io"
    kind: VirtualMachine
    name: example-vm
```

The following volumes present in the VM will be exported:

* PersistentVolumeClaims
* DataVolumes
* MemoryDump

All other volume types are not exported. To avoid the export of inconsistent data, a Virtual Machine can only be exported while it is powered off. Any active VM exports will be terminated if the Virtual Machine is started. To export data from a running Virtual Machine you must first create a Virtual Machine Snapshot (see below).

If the VM contains multiple volumes that can be exported, each volume will get its own URL links. If the VM contains no volumes that can be exported, the VMExport will go into a `Skipped` phase, and no export server is started.

### Export Virtual Machine Snapshot volumes

> After you have created the token you can now create a VMExport CR that identifies the Virtual Machine Snapshot you want to export. You can create a VMExport that looks like this:

Reviewer: Writing "After you have created the token" in each subsection seems redundant. Perhaps you could add a paragraph above the various export types saying after you have created the token, you can now create one of the following N CRs depending on the data you would like to export.

metadata:
  name: example-export
spec:
  tokenSecretRef: example-token
  source:
    apiGroup: ""
    kind: PersistentVolumeClaim
    name: example-pvc
```

In this example the PVC name is `example-pvc`. Note the PVC doesn't need to contain a Virtual Machine Disk, it can contain any content, but the main use case is exporting Virtual Machine Disks. After you post this yaml to the cluster, a new export server is created in the same namespace as the PVC. If the source PVC is *in use by another pod* (such as the virt-launcher pod) then the export will remain pending until the PVC is no longer in use. If the exporter server is active and another pod starts using the PVC, the exporter server will be terminated until the PVC is not in use anymore.

### Export status links

The VirtualMachineExport CR will contain a status with internal and external links to the export service. The internal links are only valid inside the cluster, and the external links are valid for external access through an Ingress or Route. The `cert` field will contain the CA that signed the certificate of the export server for internal links, or the CA that signed the Route or Ingress.

#### KubeVirt content-type

> Example of exporting a PVC that contains a KubeVirt disk image. The controller determines if the PVC contains a kubevirt disk by checking if there is a special annotation on the PVC, or if there is a DataVolume ownerReference on the PVC, or if the PVC has a volumeMode of block.

Reviewer: This starts with a sentence fragment; suggest "The following is an example..."

    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-06-21T14:09:02Z"
    reason: pvcBound
    status: "True"
    type: PVCReady
  links:
    external:
      cert: |-
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      volumes:
      - formats:
        - format: raw
          url: https://vmexport-proxy.test.net/api/export.kubevirt.io/v1alpha1/namespaces/example/virtualmachineexports/example-export/volumes/example-disk/disk.img
        - format: gzip
          url: https://vmexport-proxy.test.net/api/export.kubevirt.io/v1alpha1/namespaces/example/virtualmachineexports/example-export/volumes/example-disk/disk.img.gz
        name: example-disk
    internal:
      cert: |-
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      volumes:
      - formats:
        - format: raw
          url: https://virt-export-example-export.example.svc/volumes/example-disk/disk.img
        - format: gzip
          url: https://virt-export-example-export.example.svc/volumes/example-disk/disk.img.gz
        name: example-disk
  phase: Ready
  serviceName: virt-export-example-export
```
#### Archive content-type

> 

Reviewer: Could use some explanation here, does "archive content-type" apply any time the content is determined not to be a kubevirt disk image?

* Dir. A directory listing, allowing you to find the files contained in the PVC.
* Tar.gz The contents of the PVC tarred and gzipped in a single file.

Raw and Gzip will be selected if the PVC is determined to be a KubeVirt disk. KubeVirt disks contain a single disk.img file (or are a block device). Dir will return a list of the files in the PVC, to download a specific file you can replace `/dir` in the URL with the path and file name. For instance if the PVC contains the file `/example/data.txt` you can replace `/dir` with `/example/data.txt` to download just data.txt file. Or you can use the tar.gz URL to get all the contents of the PVC in a tar file.

#### Internal link certificates
The export server certificate is valid for 7 days after which it is rotated by deleting the export server pod and associated secret and generating a new one. If for whatever reason the export server pod dies, the associated secret is also automatically deleted and a new pod and secret are generated. The VirtualMachineExport object status will be automatically updated to reflect the new certificate.

#### External link certificates

> The external link certificates as associated with the Ingress/Route that points to the service created by the KubeVirt operator. The CA that signed the Ingress/Route will part of the certificates. 

Reviewer: ```suggestion
The external link certificates are associated with the Ingress/Route that points to the service created by the KubeVirt operator. The CA that signed the Ingress/Route will be part of the certificates. 
```

          url: https://virt-export-example-export.example.svc/volumes/example-dv/disk.img
        - format: gzip
          url: https://virt-export-example-export.example.svc/volumes/example-dv/disk.img.gz
        name: example-disk
  phase: Ready
  serviceName: virt-export-example-export
```
Note in this example we are in the `example` namespace in the source cluster, which is why the internal links domain ends with `.example.svc`. The external links are what will be visible to outside of the source cluster, so we can use that for when we import into the target cluster.

Now we are ready to import this disk into the target cluster. In order for CDI to import, we will need to provide it the CA certificate that signed the Ingress/Route that we will be connecting to. Luckily the `cert` field of the external links will contain the entire certificate chain. You can use the contents of the `cert` field to populate the configMap. In the target cluster create the configmap in the namespace you will import the disk into:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: router-cert
data:
  ca.pem: |
    -----BEGIN CERTIFICATE-----
    ...
    -----END CERTIFICATE-----
```

> Next create a secret in the same namespace on the target cluster, note the token value will be a header that is passed to the server by CDI.

Reviewer: ```suggestion
Next, create a secret in the same namespace on the target cluster. Note: the token value will be a header that is passed to the server by CDI.
```

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-headers
stringData:
  token: "x-kubevirt-export-token:1234567890ab"
```
Note, make sure there is no ` ` between the `:` and the actual token, otherwise the space is sent as part of the header, the authentication will fail.

> Now we can go ahead and import the disk image into the target cluster using a data volume, for convenience I put the data volume inside a data volume template section of the same VM spec as the source:

Reviewer: ```suggestion
Now we can go ahead and import the disk image into the target cluster using a data volume. For convenience, I put the data volume inside a data volume template section of the same VM spec as the source:
```

# Export API

## Prerequesites
None

### Export Feature Gate

> VMExport support must be enabled in the feature gates to be supported. The

Reviewer 2: How about:

VMExport must be enabled in the feature gates for this feature to be available

There is no "support" for any features in an open source project.

Author: Done

[feature gates](./activating_feature_gates.md#how-to-activate-a-feature-gate)
field in the KubeVirt CR must be expanded by adding the `VMExport` to it.

### Export token

> In order to securely export a Virtual Machine Disk, you must create a token that is used to authorize users accessing the export endpoint. This token must be in a secret that is stored in the namespace the Virtual Machine resides in. The contents of the secret can be passed as a token header or parameter to the export URL. The name of the header or argument is `x-kubevirt-export-token` with a value that matches the content of the secret. The secret can be named any valid secret in the namespace. We recommend you generate an alpha numeric token of at least 12 characters. The data key should be `token`. For example:

Reviewer 2: in the namespace where the Virtual Machine resides.

Author: Done

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: example-token
data:
  token: MTIzNDU2Nzg5MGFi
```
### Export Virtual Machine volumes
After you have created the token you can now create a VMExport CR that identifies the Virtual Machine you want to export. You can create a VMExport that looks like this:

```yaml
apiVersion: export.kubevirt.io/v1alpha1
kind: VirtualMachineExport
metadata:
  name: example-export
spec:
  tokenSecretRef: example-token
  source:
    apiGroup: "kubevirt.io"
    kind: VirtualMachine
    name: example-vm
```

The following volumes present in the VM will be exported:

* PersistentVolumeClaims
* DataVolumes
* MemoryDump

> All other volume types are not exported. The export will only be available when the VM is not running, this is to prevent corrupt data from being exported. If you want to export data from a running VM, the VM will have to use storage that can be snapshotted, and you can create a Virtual Machine Snapshot and export that while the VM stays running. If an export is active and the VM is started the export will detect this state and terminate the export server.

Reviewer 2: To avoid the export of inconsistent data, a Virtual Machine can only be exported while it is powered off.  Any active VM exports will be terminated if the Virtual Machine is started.  To export data from a running Virtual Machine you must first create a Virtual Machine Snapshot (see below).

Author: Done

  - lastProbeTime: null
    lastTransitionTime: "2022-06-21T14:09:02Z"
    reason: pvcBound
    status: "True"
    type: PVCReady
  links:
    external:
      cert: |-
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      volumes:
      - formats:
        - format: dir
          url: https://vmexport-proxy.test.net/api/export.kubevirt.io/v1alpha1/namespaces/example/virtualmachineexports/example-export/volumes/example/dir
        - format: tar.gz
          url: https://vmexport-proxy.test.net/api/export.kubevirt.io/v1alpha1/namespaces/example/virtualmachineexports/example-export/volumes/example/disk.tar.gz
        name: example-disk
    internal:
      cert: |-
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      volumes:
      - formats:
        - format: dir
          url: https://virt-export-example-export.example.svc/volumes/example/dir
        - format: tar.gz
          url: https://virt-export-example-export.example.svc/volumes/example/disk.tar.gz
        name: example-disk
  phase: Ready
  serviceName: virt-export-example-export
```
#### Format types
There are 4 format types that are possible:

> * Raw. The unaltered raw KubeVirt disk image.

Reviewer: Looking at the rendered copy, the following bullets did not render properly because there has to be a newline before a list.

```suggestion

* Raw. The unaltered raw KubeVirt disk image.
```

Author: Weird my local renderer renders it properly. Added the newline.

# Export API

## Prerequesites

> None

Reviewer 3: why None if the feature gate must be enabled? isnt it a Prerequesite?

Author: Because featuregates are a separate item in the documentation. So besides the featuregate, there are none.

Reviewer 4: Yeah I think I'm with @ShellyKa13.  Seems like exactly the sort of thing that should go here.

Author: Well all the other pages have separate sections for enabling the feature gate. I think I am going to just remove the prerequesites section since it makes no sense.

### Export Feature Gate

VMExport support must be enabled in the feature gates to be supported. The
[feature gates](./activating_feature_gates.md#how-to-activate-a-feature-gate)
field in the KubeVirt CR must be expanded by adding the `VMExport` to it.

### Export token

> In order to securely export a Virtual Machine Disk, you must create a token that is used to authorize users accessing the export endpoint. This token must be in a secret that is stored in the namespace the Virtual Machine resides in. The contents of the secret can be passed as a token header or parameter to the export URL. The name of the header or argument is `x-kubevirt-export-token` with a value that matches the content of the secret. The secret can be named any valid secret in the namespace. We recommend you use best security practices when generating this token. 

Reviewer 3: maybe add a yaml for example?

Author: Its just creating a secret, the 'hard' part is the generation of the token itself. I can add a yaml to create a secret but it is not very interesting.

### Export Virtual Machine disk

After you have created the token you can now create a VMExport CR that identifies the disk of the VM you want to export. In KubeVirt in the end all persisted Virtual Machine disks are stored in a Persistent Volume Claim (PVC) in the namespace the Virtual Machine resides in. Once you have identified the PVC you can create a VMExport that looks like this:

```yaml
apiVersion: export.kubevirt.io/v1alpha1
kind: VirtualMachineExport
metadata:

>     name: example-export

Author: Done

spec:
  tokenSecretRef: example-token
  source:
    apiGroup: ""
    kind: PersistentVolumeClaim
    name: vm-disk
```

> In this example the PVC name is vm-disk, but it will be different for your Virtual Machine Disk. Note the PVC doesn't need to contain a Virtual Machine Disk, it can contain any content, but the main use case is exporting Virtual Machine Disks. After you post this yaml to the cluster, and the PVC is not in use by another pod, a new export server is created in the same namespace as the PVC. The VirtualMachineExport CR will contain a status with internal and external links to the export service. The internal links are only valid inside the cluster, and the external links are valid for external access. The `cert` field will contain the CA that signed the certificate of the export server for internal links, or the CA that signed the Route or Ingress.

Reviewer 3: maybe hightlight **not in use by another pod**

        ...
        -----END CERTIFICATE-----
      volumes:
      - formats:
        - format: raw
          url: https://vmexport-proxy.test.net/api/export.kubevirt.io/v1alpha1/namespaces/example/virtualmachineexports/example-export/volumes/example-disk/disk.img
        - format: gzip
          url: https://vmexport-proxy.test.net/api/export.kubevirt.io/v1alpha1/namespaces/example/virtualmachineexports/example-export/volumes/example-disk/disk.img.gz
        name: example-disk
    internal:
      cert: |-
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
        -----BEGIN CERTIFICATE-----
        ...
        -----END CERTIFICATE-----
      volumes:
      - formats:
        - format: raw
          url: https://virt-export-example-export.example.svc/volumes/example-disk/disk.img
        - format: gzip
          url: https://virt-export-example-export.example.svc/volumes/example-disk/disk.img.gz
        name: example-disk
  phase: Ready
  serviceName: virt-export-example-export
```

If the source PVC is in use by another pod (such as the virt-launcher pod) then the export will remain pending until the PVC is no longer in use.

> 

Reviewer 3: An example for how to use the created links is welcomed, maybe aexample to use it as source for new dv? and download to laptop?

Author: I added an example of how to copy an image from one cluster to another using CDI.
